Entrypoint bugfixes

antarcticrainforest opened the pull request:
This PR mainly aims at using the new conda-forge package for the freva-rest-server.

Other changes were done to make the type checks work.

antarcticrainforest commented:
FYI the tests fail because the test coverage is lower as some the try except blocks aren't fully covered anymore. I didn't look into as I don't think this is important at all.

mo-dkrz commented on freva-rest/Dockerfile:
```diff
@@ -1,110 +1,39 @@
-    API_REDIS_SSL_CERTFILE=/certs/client-cert.pem \
-    API_REDIS_SSL_KEYFILE=/certs/client-key.pem \
-    API_MONGO_USER=mongo \
-    API_MONGO_PASSWORD=secret \
```
since we remove those variable here, I think it would be better to have some default values for each in the `init_mongo` in conda recepie. Because otherwise the user runs into the [unbounded variable issue](https://gitlab.dkrz.de/freva/metadata-crawler-source/-/jobs/625400). Or if you think those variables have to be mandatory and there is not need to define the default for them, imho we need to validate the environment variables of users in the entrypoint level to give the better logs and suggestion to user when it doesn't define proper env vars. Otherwise user feels like our produced image is broken and not working and the rest of story ...! what do you think?

mo-dkrz commented on freva-rest/entrypoint.sh:
```diff
@@ -50,102 +45,45 @@ log_service() {
+    log_service "Starting solr service"
+    /opt/conda/bin/solr start -force
+    SOLR_PORT=${API_SOLR_PORT:-8983}
+    until curl -s "http://localhost:${SOLR_PORT}/solr/admin/ping" >/dev/null 2>&1; do
```
could you please install `curl` in `Dockerfile` first, otherwise image-run stops at this point since `curl` is not installed in mamba image by default, OR if you think this validation here is not important you can simply remove it!

antarcticrainforest replied:
Done!

mo-dkrz reviewed (changes requested):
Two important comments:

- Installing the freva-data-portal dependencies in container or conda level
- Making a decision about mongo username and pass default value to not run into the unbounded variabel issue

Apart from that, could you please furnish the README of the freva-rest for newly added `USE_MYSQL` and `USE_REDIS`?

antarcticrainforest commented on freva-rest/Dockerfile:
```diff
@@ -1,110 +1,39 @@
 COPY . .
-RUN python3 -m pip install --upgrade pip --break-system-packages && \
-    python3 -m pip install build --break-system-packages && \
-    python3 -m pip install . --break-system-packages && \
```
I ran a pip check and added cloudpickle to the dependencies. Good catch!

antarcticrainforest commented on .github/workflows/build_job.yml:
```diff
@@ -121,6 +121,7 @@ jobs:
           build-args: VERSION=${{ needs.release-type-determination.outputs.tag }}
           platforms: linux/amd64, linux/arm64
```
ok.

mo-dkrz reviewed (changes requested):
Here is the produced image test run result:

Apart from fixing the issue with Solr, for others we need to furnish the README!

1. Generally for running the built image, defining the `API_OIDC_DISCOVERY_URL` is required. So I think it needs to be mentioned in the REAMDE as **required** env.

2. Enabling Redis works perfect and doesn't require any environment or doesn't stop the process, but enabling Mongo requires `API_MONGO_PASSWORD` and `API_MONGO_USER`. So I think these two env vars also have to be mentioned as *required* in README. 

```bash
docker run -e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -e API_MONGO_USER=mongo -e API_MONGO_PASSWORD=secret -e USE_MONGODB=1 -p 
7777:7777 ghcr.io/freva-clint/freva-rest-api:2502.0.1-dev1 python3 -m freva_rest.cli
```

3. Enabling the MYSQL:
It requires defining `MYSQL_ROOT_PASSWORD`, `MYSQL_USER`, `MYSQL_PASSWORD` , and `MYSQL_DATABASE` beforehand. So these are the ones we need to mention as *required* somewhere to announce user before running.

```bash
docker run -e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -e API_MONGO_USER=mongo -e API_MONGO_PASSWORD=secret -e USE_MYSQL=1 -e MYSQL_ROOT_PASSWORD=pass -e MYSQL_USER=user -e MYSQL_PASSWORD=pass -e MYSQL_DATABASE=db -p 7777:7777 ghcr.io/freva-clint/freva-rest-api:2502.0.1-dev1 python3 -m freva_rest.cli
```

But the most important and main issue with the MySQL is [here](https://github.com/conda-forge/freva-rest-server-feedstock/blob/main/recipe/build.sh#L59) which leads to stop running the image:

```bash
2025-02-18T15:30:04.554435Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
2025-02-18T15:30:04.643291Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
2025-02-18T15:30:04.773831Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
2025-02-18T15:30:04.773854Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
2025-02-18T15:30:04.867486Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Socket: /tmp/mysqlx.sock
2025-02-18T15:30:04.867537Z 0 [System] [MY-010931] [Server] /opt/conda/bin/mysqld: ready for connections. Version: '9.0.1'  socket: '/tmp/mysql.sock'  port: 0  conda-forge.
Shutting down MySQL
.2025-02-18T15:30:09.416907Z 0 [System] [MY-013172] [Server] Received SHUTDOWN from user <via user signal>. Shutting down mysqld (Version: 9.0.1).
2025-02-18T15:30:09.745642Z 0 [System] [MY-010910] [Server] /opt/conda/bin/mysqld: Shutdown complete (mysqld 9.0.1)  conda-forge.
2025-02-18T15:30:09.745670Z 0 [System] [MY-015016] [Server] MySQL Server - end.
 * 
/opt/conda/libexec/freva-rest-server/scripts/init-mysql: line 41: kill: (53) - No such process
 * MySQL server PID file could not be found!
```

As I see in the logs, it can properly shutdown MYSQL, so I'm not sure about the necessity of this termination command there!

4. Enabling the Solr (important one):

When we enable Solr, it doesn't matter what port we determine before running, we always get `the Port <port-number> is already being used by another process` and since it's an error, it will be stuck in the [Solr stop trap](https://github.com/conda-forge/freva-rest-server-feedstock/blob/main/recipe/build.sh#L74) and then it stops running the rest of entrypoint routine!

To reproduce the error, I executed the image and ran the `/bin/bash /opt/conda/libexec/freva-rest-server/scripts/init-solr`. when I run inside if the running image, it's not stuck in trap, but when it goes through [the latest stop Solr](https://github.com/conda-forge/freva-rest-server-feedstock/blob/main/recipe/build.sh#L87-L89) in build recipe, it makes an exit and leads to stop running the image!

```bash
docker run -e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -e USE_SOLR=1 -e API_SOLR_PORT=8988 -p 7777:7777 ghcr.io/freva-clint/fr
eva-rest-api:2502.0.1-dev1 python3 -m freva_rest.cli
```
logs:
```bash
[2025-02-18 15:36:44] [SERVICE] === Initialising Solr ===
Java 23 detected. Enabled workaround for SOLR-16463
Waiting up to 180 seconds to see Solr running on port 8988 [|]  
Started Solr server on port 8988 (pid=65). Happy searching!

WARNING: Using _default configset. Data driven schema functionality is enabled by default, which is
         NOT RECOMMENDED for production use.

         To turn it off:
            curl http://localhost:8983/solr/files/config -d '{"set-user-property": {"update.autoCreateFields":"false"}}'
         Or:
            bin/solr config -c files -s http://localhost:8983 --action set-user-property --property update.autoCreateFields --value false

Created new core 'files'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0{
  "responseHeader":{
    "status":0,
    "QTime":76
  },
  "WARNING":"This response format is experimental.  It is likely to change in the future."
100   209  100   151  100    58   1662    638 --:--:-- --:--:-- --:--:--  2322

Port 8988 is already being used by another process (pid: 65)
Please choose a different port using the -p option.

Sending stop command to Solr running on port 8988 ... waiting up to 180 seconds to allow Jetty process 65 to stop gracefully.
```

antarcticrainforest commented:
Thanks a lot for the thorough comments. I should have addressed your raised points. Either here or by fixing the according setup script in the conda-forge repo. Here are some sanity checks:

```console
$: docker run -v ./entrypoint.sh://docker-entrypoint-initdb.d/entrypoint.sh -e USE_SOLR=1 -e USE_MONGODB=1 -e \ 
USE_REDIS=1 -e USE_MYSQL=1 -v $PWD/init-solr:/opt/conda/libexec/freva-rest-server/scripts/init-solr \ 
-v $PWD/init-mysql:/opt/conda/libexec/freva-rest-server/scripts/init-mysql \
-e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -v \
$PWD/init-mongo:/opt/conda/libexec/freva-rest-server/scripts/init-mongo -e API_MONGO_USER=mongo -e \
API_MONGO_PASSWORD=secret -e MYSQL_PASSWORD=secret -e MYSQL_ROOT_PASSWORD=foo -e \
MYSQL_USER=freva -e MYSQL_DATABASE=freva --net=host -it freva-rest:latest --dev --debug
```
- Freva rest api
 ```console
 $: freva-client databrowser metadata-search --host localhost:7777 --json|jq -rc 'keys'
    ["ensemble","experiment","institute","model","product","project","realm","time_aggregation","time_frequency","variable"]
# No error/warnings on the frest-api
```
- Reids
```console
$: python -c 'import redis; print(redis.Redis(host="localhost", port=6379, db=0).ping())'
   True
```
- Mysql 
```console
$: mariadb --ssl=0 -u freva -p'secret' -h localhost -P 3306 -D freva -e "SHOW TABLES;" 
+--------------------------------------+
| Tables_in_freva                      |
+--------------------------------------+
| auth_group                           |
| auth_group_permissions               |
| auth_permission                      |
| auth_user                            |
| auth_user_groups                     |
| auth_user_user_permissions           |
| django_admin_log                     |
| django_flatpage                      |
| django_flatpage_sites                |
| django_migrations                    |
| django_session                       |
| django_site                          |
| externaluser_externaluser            |
| future_files                         |
| future_recipe                        |
| hindcast_frontend_hindcastevaluation |
| history_batch_settings               |
| history_configuration                |
| history_history                      |
| history_historytag                   |
| history_output                       |
| history_result                       |
| history_resulttag                    |
| plugins_parameter                    |
| plugins_toolpullrequest              |
| plugins_user                         |
| plugins_version                      |
| solr_usercrawl                       |
+--------------------------------------+

```


I  didn't check authentication as our beloved sys admin seemed to have fudged all systems.

mo-dkrz reviewed (commented):
Great job @antarcticrainforest !

As you already mentioned, now it works properly:
```bash
docker run -e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -e USE_MONGODB=1 -e API_MONGO_USER=mongo -e API_MONGO_PASSWORD=secret -e USE_SOLR=1 -e USE_REDIS=1 -e USE_MYSQL=1 -e MYSQL_ROOT_PASSWORD=pass -e MYSQL_USER=user -e MYSQL_PASSWORD=pass -e MYSQL_DATABASE=db -p 7777:7777 ghcr.io/freva-clint/freva-rest-api:2502.0.1-dev1 python3 -m freva_rest.cli
```
However my only concern is mentioning the following environment variable as *required* somewhere in README or Documentation or where-ever you feel alright, to get a better experience to our future users:

1. API_OIDC_DISCOVERY_URL
2. API_MONGO_USER (if USE_MONGODB is enabled)
3. API_MONGO_PASSWORD (if USE_MONGODB is enabled)
4. MYSQL_ROOT_PASSWORD (if USE_MYSQL is enabled)
5. MYSQL_USER (if USE_MYSQL is enabled)
6. MYSQL_PASSWORD (if USE_MYSQL is enabled)
7. MYSQL_DATABASE (if USE_MYSQL is enabled)

Apart from that, LGTM!

mo-dkrz commented on freva-rest/README.md:
```diff
@@ -49,13 +49,20 @@ The container supports several startup modes on `AMD64` and `ARM64`:
 
```
Could you please add `ppc64le` here as well?

antarcticrainforest replied:
Didn't I do that already?

mo-dkrz replied:
Updating in REAME I meant.

antarcticrainforest replied:
I don't think that we need to mention the architectures at all. Docker will pull the right one for the machine, if it isn't available then it'll simply fail.

antarcticrainforest commented:
> Great job @antarcticrainforest !
> 
> As you already mentioned, now it works properly:
> ```bash
> docker run -e API_OIDC_DISCOVERY_URL=https://freva-keycloak.cloud.dkrz.de/realms/Freva/.well-known/openid-configuration -e USE_MONGODB=1 -e API_MONGO_USER=mongo -e API_MONGO_PASSWORD=secret -e USE_SOLR=1 -e USE_REDIS=1 -e USE_MYSQL=1 -e MYSQL_ROOT_PASSWORD=pass -e MYSQL_USER=user -e MYSQL_PASSWORD=pass -e MYSQL_DATABASE=db -p 7777:7777 ghcr.io/freva-clint/freva-rest-api:2502.0.1-dev1 python3 -m freva_rest.cli
> ```
> However my only concern is mentioning the following environment variable as *required* somewhere in README or Documentation or where-ever you feel alright, to get a better experience to our future users:
> 
> 1. API_OIDC_DISCOVERY_URL
> 2. API_MONGO_USER (if USE_MONGODB is enabled)
> 3. API_MONGO_PASSWORD (if USE_MONGODB is enabled)
> 4. MYSQL_ROOT_PASSWORD (if USE_MYSQL is enabled)
> 5. MYSQL_USER (if USE_MYSQL is enabled)
> 6. MYSQL_PASSWORD (if USE_MYSQL is enabled)
> 7. MYSQL_DATABASE (if USE_MYSQL is enabled)
> 
> Apart from that, LGTM!

Technically they only become mandatory if you want to use the service. I figured that by rather than mentioning something in a readme noone reads I make the entrypoint script with a meaningful error message.

You can check what happens if you start the container without any Env variable and use one of the services.

mo-dkrz commented:
Since you are on README, I have another point here. If you look at the packages page, you see the REAME of the main directory of repo, while having the freva-rest README makes more sense there! https://github.com/FREVA-CLINT/freva-nextgen/pkgs/container/freva-rest-api

What do you think of overriding the description of package after build and push stage in ci build action via github api?

antarcticrainforest commented:
@mo-dkrz I should have addressed all points in one way or the other. Except for the overriding of the readme. I am not sure whether or not this is possible. Because you assign a package to a repository, once you have done that github will automatically use the main readme of the repository. This is totally unrelated to the fact that this is a docker image.

mo-dkrz reviewed (approved):
LGTM!
